packages/versioning/src/lib.rs

```diff
@@ -1,115 +1,183 @@
-        Self::Query { query }
-    }
-
-    pub fn handle_query<const VERSION: Version, F>(self, f: F) -> StdResult<Binary>
+pub fn parse_semver(version: &str) -> SemVer {
+    fn parse_segment<'r, I>(
+        iter: &mut I,
+        lowercase_name: &str,
+        pascal_case_name: &str,
+    ) -> VersionSegment
     where
-        F: FnOnce(Q) -> StdResult<Binary>,
+        I: Iterator<Item = &'r str> + ?Sized,
     {
-        match self {
-            WithVersion::Version { version: () } => to_binary(&VERSION),
-            WithVersion::Query { query } => f(query),
-        }
+        iter.next()
+            .unwrap_or_else(|| panic!("No {lowercase_name} segment in version string!"))
+            .parse()
+            .unwrap_or_else(|_| {
+                panic!("{pascal_case_name} segment in version string is not a number!")
+            })
+    }
+
+    let mut iter = version.split('.');
+
+    let major: VersionSegment = parse_segment(&mut iter, "major", "Major");
+    let minor: VersionSegment = parse_segment(&mut iter, "minor", "Minor");
+    let patch: VersionSegment = parse_segment(&mut iter, "patch", "Patch");
+
+    if iter.next().is_some() {
+        panic!("Unexpected fourth segment found in version string!");
+    };
+
+    SemVer {
+        major,
+        minor,
+        patch,
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use serde::{Deserialize, Serialize};
+#[macro_export]
+macro_rules! package_version {
+    () => {{
+        $crate::parse_semver(::core::env!(
+            "CARGO_PKG_VERSION",
+            "Cargo package version is not set as an environment variable!",
+        ))
+    }};
+}
 
-    use super::WithVersion;
+const VERSION_PAIR_ITEM: Item<'static, VersionPair> = Item::new("contract_version_pair");
+
+pub fn initialize<const STORAGE_VERSION: VersionSegment>(
+    storage: &mut dyn Storage,
+    component_version: SemVer,
+) -> StdResult<()> {
+    VERSION_PAIR_ITEM.save(
+        storage,
+        &VersionPair {
+            storage: STORAGE_VERSION,
+            software: component_version,
+        },
+    )
+}
 
-    #[derive(Debug, Eq, PartialEq, Serialize, Deserialize)]
-    #[serde(rename_all = "snake_case")]
-    enum VariantsQuery {
-        Abc {},
-        Def {},
-        Version {},
+// TODO remove when all contracts have been migrated to post-refactor versions
+pub fn upgrade_old_contract<
+    'r,
+    const OLD_COMPATIBILITY_VERSION: VersionSegment,
+    MigrateStorageFunctor,
+    MigrateStorageError,
+>(
+    storage: &'r mut dyn Storage,
+    component_version: SemVer,
+    migrate_storage_functor: Option<MigrateStorageFunctor>,
+) -> Result<(), MigrateStorageError>
+where
+    MigrateStorageFunctor: FnOnce(&'r mut dyn Storage) -> Result<(), MigrateStorageError>,
+    MigrateStorageError: From<StdError> + Error,
+{
+    pub const CW_VERSION_ITEM: Item<'static, u16> = Item::new("contract_info");
+
+    pub const OLD_VERSION_ITEM: Item<'static, u16> = Item::new("contract_version");
```

Comment:
do we need them `public`?

Reply:
No, that's a, while not an issue, a mistake.